src/lib/sync_handler/sync_handler.ml

```diff
@@ -36,15 +36,24 @@ module Make (Inputs : Inputs_intf) :
    and type syncable_ledger_answer := Sync_ledger.answer = struct
   open Inputs
 
+  let get_breadcrumb_ledgers frontier =
+    List.map
+      (Transition_frontier.all_breadcrumbs frontier)
+      ~f:
+        (Fn.compose Staged_ledger.ledger
+           Transition_frontier.Breadcrumb.staged_ledger)
+
   let get_ledger_by_hash ~frontier ledger_hash =
-    List.find_map (Transition_frontier.all_breadcrumbs frontier) ~f:(fun b ->
-        let ledger =
-          Transition_frontier.Breadcrumb.staged_ledger b
-          |> Staged_ledger.ledger
-        in
-        if Ledger_hash.equal (Ledger.merkle_root ledger) ledger_hash then
-          Some ledger
-        else None )
+    let ledger_breadcrumbs =
+      Sequence.of_lazy
+        (lazy (Sequence.of_list @@ get_breadcrumb_ledgers frontier))
+    in
+    Sequence.append
+      (Sequence.singleton
+         (Transition_frontier.shallow_copy_root_snarked_ledger frontier))
+      ledger_breadcrumbs
+    |> Sequence.find ~f:(fun ledger ->
+           Ledger_hash.equal (Ledger.merkle_root ledger) ledger_hash )
```

Comment:
No need for sequence here if all breadcrumbs is a list. The lazy here doesn't seem to buy us anything, am I missing something?

Reply:
`all_breadcrumbs` takes O(n) time to construct a list of Breadcrumbs. Namely, because we are map all the values of the hashtable `transition_frontier.table` into Breadcrumbs.  We mainly want to sync with a peer's snarked ledger and if the snarked ledger doesn't match with the ledger that we look for, we go through the ledgers in their `transition_frontier`. So the laziness does save us time. I would predict that 99.9% of the time, this statement will take O(1) instead of O(n).

Reply:
yeah but get_breadcrumb_ledgers already gives you a full list in O(n) right?

Reply:
That is true. Having this lazy evaluation did make the performance of my test improve significantly faster. So, we should keep this.

Reply:
By the way, with the `lazy` keyword. The following state will not get printed
```ocaml
let hi = lazy (printf "Hi, my name is John");;
```

However, when I do `Lazy.force hi`, it gets evaluated and the statement gets printed.

Reply:
Talked with John, the lazy isn't forced whenever `shallow_copy_root_snarked_ledger` emits something, so this is a time saver.